### PR TITLE
Added 404 and 500 error pages

### PR DESCRIPTION
**Pull request by dcrystalj**

### CODE REVIEW

**Comment by mitar** [comment]
Great, but like everything, we will also do this in a bit more precise way than just quick way. Are you ready? Is this OK with you? ;-)

**Comment by dcrystalj** [comment]
do you mean writing own 404 view ?. but i don't see any reason writing it.

**Comment by mitar** [comment]
So there are in fact four things you want to cover:
- 404
- 403
- 500
- CSRF error

So read [this](https://docs.djangoproject.com/en/dev/topics/http/urls/#error-handling) and [this](https://docs.djangoproject.com/en/dev/ref/contrib/csrf/#csrf-failure-view). For CSRF we can use same view as for 500, because by default CSRF template is hard-coded.

So what I would do:
- define handlers in `urls.py` so that it is obvious what is happening, for 404 and 500 you can use default views, for 403 and CSRF we define our common view combining both
- in `urls.py` define debug URLs for seeing how 403/404/500 views look like
- configure CSRF view in `settings.py`
- define all three templates

I suggest that you check how things are done in [nodewatcher](https://github.com/wlanslovenija/nodewatcher). ;-)

File: piplmesh/frontend/templates/500.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+    <head>
+        <title>500 Internal Server Error – Piplmesh, wlan slovenija</title>
+        <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
+        <meta name="robots" content="NONE,NOARCHIVE" />
+    </head>
+    <body>
+        <h2>500 Internal Server Error</h2>
+        <p>An internal server error has ocurred.</p>
```
Comment (mitar) [inline piplmesh/frontend/templates/500.html]: OK, I see that you have been already using nodewatcher templates. Here is spelling mistake I just noticed. :-) There is double _c_ in _occured_. :-)

File: piplmesh/frontend/templates/500.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+    <head>
+        <title>500 Internal Server Error – Piplmesh, wlan slovenija</title>
```
Comment (mitar) [inline piplmesh/frontend/templates/500.html]: I would just leave title as it is, `500 Internal Server Error`.

File: piplmesh/frontend/templates/404.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{% extends "plain.html" %}
```
Comment (mitar) [inline piplmesh/frontend/templates/404.html:1]: Nice, simple template. Good.

File: piplmesh/frontend/templates/404.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+{% load i18n %}
+
+{% block title %}{% trans "Error page" %}{% endblock %}
```
Comment (mitar) [inline piplmesh/frontend/templates/404.html]: I would use same title here: _404 Not Found_.

**Comment by mitar** [comment]
It looks good. so add also 403 template please and as described above. Otherwise really a nice small pull request. Congrats.

**Comment by dcrystalj** [comment]
please comment. Still not sure if i did csrf right

File: piplmesh/settings.py [hunk 5]
```diff
@@ -278,3 +278,5 @@
 # you will be explicitly warned that you have to change the code to take effect, before you will make the change.
 # Current settings are autocomplete, searching whole web.
 SEARCH_ENGINE_UNIQUE_ID = '003912915932446183218:zeq20qye9oa'
+
+CSRF_FAILURE_VIEW = 'frontend.views.ForbiddenView'
```
Comment (mitar) [inline piplmesh/settings.py]: `piplmesh.frontend.views.ForbiddenView`

File: piplmesh/urls.py [hunk 6]
```diff
@@ -38,3 +38,14 @@
...
     url(r'^api/', include(v1_api.urls)),
 )
+
+handler403 = frontend_views.ForbiddenView
+handler404 = 'django.views.defaults.page_not_found'
+handler500 = 'django.views.defaults.server_error'
+
+if getattr(settings, 'DEBUG', False):
+    urlpatterns += patterns('',
+    (r'^403/$', handler403),
```
Comment (mitar) [inline piplmesh/urls.py]: Good, but indent once more those three lines.

File: piplmesh/urls.py [hunk 7]
```diff
@@ -1,5 +1,5 @@
 from django.conf.urls.defaults import patterns, include, url
-
+from django.conf import settings
```
Comment (mitar) [inline piplmesh/urls.py:1]: Move this import to the top: we group imports by package and sort them alphabetically.

File: piplmesh/frontend/templates/404.html [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{% extends "plain.html" %}
+
+{% load i18n %}
+
+{% block title %}{% trans "404 Not Found" %}{% endblock %}
```
Comment (mitar) [inline piplmesh/frontend/templates/404.html:5]: Add probably also:

```
{% block extrahead %}<meta name="robots" content="NONE,NOARCHIVE" />{% endblock %}
```

File: piplmesh/frontend/templates/403.html [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{% extends "plain.html" %}
+
+{% load i18n %}
+
+{% block title %}{% trans "403 Forbidden" %}{% endblock %}
```
Comment (mitar) [inline piplmesh/frontend/templates/403.html:5]: Same here:

```
{% block extrahead %}<meta name="robots" content="NONE,NOARCHIVE" />{% endblock %}
```

File: piplmesh/frontend/templates/403.html [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+{% block content %}
+<h2>{% trans "403 Forbidden" %}</h2>
+
+<div class="error">
+    <p>{% trans "You don't have permission to access the requested page." %}</p>
```
Comment (mitar) [inline piplmesh/frontend/templates/403.html]: _You don't have permission to access the page you requested._

File: piplmesh/frontend/views.py [hunk 11]
```diff
@@ -1,7 +1,8 @@
 from django.views import generic as generic_views
-
+from django.http import HttpResponseForbidden
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Import only modules. Not its content. See how we are doing it elsewhere. (The only exception is `urls.py`.)

File: piplmesh/frontend/views.py [hunk 12]
```diff
@@ -1,7 +1,8 @@
...
+from django.template import RequestContext, Context, loader
+from django.conf import settings
```
Comment (mitar) [inline piplmesh/frontend/views.py:2]: Group imports, order alphabetically.

File: piplmesh/frontend/views.py [hunk 13]
```diff
@@ -26,3 +27,16 @@ class UserView(detail.DetailView):
     document = models.User
     slug_field = 'username'
     slug_url_kwarg = 'username'
+	
+def ForbiddenView(request, reason=""):
```
Comment (mitar) [inline piplmesh/frontend/views.py]: If it is not a class-based view, name it as a function, now class (don't use camel case).

File: piplmesh/frontend/views.py [hunk 14]
```diff
@@ -26,3 +27,16 @@ class UserView(detail.DetailView):
...
+    """
+    Displays 403 fobidden page. For example, when request fails CSRF protection.
+    """
+
+    from django.middleware.csrf import REASON_NO_REFERER
+    t = loader.get_template("403.html")
+    return HttpResponseForbidden(t.render(RequestContext(request, {
+        'DEBUG'  : settings.DEBUG,
```
Comment (mitar) [inline piplmesh/frontend/views.py]: No space before `:`. (nodewatcher has a bit different code style.)

File: piplmesh/frontend/templates/403.html (hunk 10 above)
Comment (mitar) [inline piplmesh/frontend/templates/403.html]: This is good, but use also other template variables you are passing to the template.

```
<div class="error">
    <p>{% trans "You don't have permission to access the page you requested." %}</p>
    {% if no_referer %}
        ... (translatable!)
    {% else %}
        <p>{% trans "For accessing this site and using all features cookies in your browser must be enabled." %}</p>
    {% endif %}
</div>

{% if DEBUG and reason %}<p>Reason: {{ reason }}</p>{% endif %}
```

**Comment by dcrystalj** [comment]
please take a look.

File: piplmesh/frontend/views.py [hunk 15]
```diff
@@ -1,3 +1,6 @@
+from django import http as http_views
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Those are not views. :-) Just leave `http` as it is. ;-)

File: piplmesh/frontend/views.py [hunk 16]
```diff
@@ -1,3 +1,6 @@
...
+from django.conf import settings
+from django.template import RequestContext, Context, loader
```
Comment (mitar) [inline piplmesh/frontend/views.py]: You are not importing only module here.

File: piplmesh/frontend/views.py [hunk 17]
```diff
@@ -26,3 +29,16 @@ class UserView(detail.DetailView):
     document = models.User
     slug_field = 'username'
     slug_url_kwarg = 'username'
+	
+def forbidden_view(request, reason=""):
+    """
+    Displays 403 fobidden page. For example, when request fails CSRF protection.
+    """
+
+    from django.middleware.csrf import REASON_NO_REFERER
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Maybe also here import only module.

File: piplmesh/frontend/templates/403.html [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+{% block extrahead %}<meta name="robots" content="NONE,NOARCHIVE" />{% endblock %}
+
+{% block content %}
+<h2>{% trans "403 Forbidden" %}</h2>
+
+<div class="error">
+    <p>{% trans "You don't have permission to access the page you requested." %}</p>
+    {% if no_referer %}
+        <p>{% trans "You are seeing this message because this HTTPS site requires a 'Referer header' to be sent by your Web browser, but none was sent. This header is required for security reasons, to ensure that your browser is not being hijacked by third parties." %}</p>
```
Comment (mitar) [inline piplmesh/frontend/templates/403.html]: Good. Just please use `blocktrans` tags for such long passages of text.

File: piplmesh/frontend/views.py [hunk 19]
```diff
@@ -26,3 +29,16 @@ class UserView(detail.DetailView):
...
+    t = loader.get_template("403.html")
+    return http_views.HttpResponseForbidden(t.render(RequestContext(request, {
+        'DEBUG': settings.DEBUG,
+        'reason': reason,
+        'no_referer' : reason == REASON_NO_REFERER,
```
Comment (mitar) [inline piplmesh/frontend/views.py]: One space too much, still.

File: piplmesh/frontend/views.py [hunk 20]
```diff
@@ -26,3 +29,16 @@ class UserView(detail.DetailView):
...
     slug_field = 'username'
     slug_url_kwarg = 'username'
+	
+def forbidden_view(request, reason=""):
+    """
+    Displays 403 fobidden page. For example, when request fails CSRF protection.
+    """
+
+    from django.middleware.csrf import REASON_NO_REFERER
+    t = loader.get_template("403.html")
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Constant string.

File: piplmesh/frontend/views.py [hunk 21]
```diff
@@ -26,3 +29,16 @@ class UserView(detail.DetailView):
     document = models.User
     slug_field = 'username'
     slug_url_kwarg = 'username'
+	
+def forbidden_view(request, reason=""):
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Constant string.

**Comment by mitar** [comment]
Otherwise it looks good.

File: piplmesh/frontend/views.py (hunk 21 above)
Comment (dcrystalj) [inline piplmesh/frontend/views.py]: sorry i don't get it. What is here constant string?

Comment (mitar) [inline piplmesh/frontend/views.py]: Empty string is a constant string, no? ;-) So it should be `''`.

**Comment by dcrystalj** [comment]
updated

File: piplmesh/frontend/views.py (hunk 21 above)
Comment (mitar) [inline piplmesh/frontend/views.py]: TODO?

Comment (dcrystalj) [inline piplmesh/frontend/views.py]: ohhh, I totally forgot

**Comment by mitar** [comment]
Great! Good job.

**Pull request merged**